File: modules/common/ipc/examples/sub_example.cpp
//=================================================================================================
// Copyright (C) 2025 GRAPE Contributors
//=================================================================================================

#include <print>
#include <thread>

#include "grape/exception.h"
#include "grape/ipc/session.h"
#include "grape/ipc/subscriber.h"
#include "topic_example.h"

//=================================================================================================
// Demonstrates a basic IPC subscriber. See pub_example.cpp for the corresponding publisher.
auto main() -> int {
  try {
    grape::ipc::init(grape::ipc::Config{});

    const auto data_cb = [](const std::expected<std::string, grape::ipc::Error>& data,
                            const grape::ipc::SampleInfo& info) -> void {
      if (data) {
        std::println("Received message: '{}' (from {})", *data, toString(info.publisher));
      } else {
        std::println("Error: {} (data from {})", toString(data.error()), toString(info.publisher));
      }
    };

    const auto match_cb = [](const grape::ipc::Match& match) -> void {
      std::println("\n{} (entity: {})", toString(match.status), toString(match.remote_entity));
    };

    const auto topic = grape::ipc::ex::ExampleTopicAttributes{};
    auto subscriber = grape::ipc::Subscriber(topic, data_cb, match_cb);

    constexpr auto SLEEP_TIME = std::chrono::milliseconds(500);
    while (grape::ipc::ok()) {
      std::this_thread::sleep_for(SLEEP_TIME);
    }
    return EXIT_SUCCESS;
  } catch (...) {
    grape::Exception::print();
    return EXIT_FAILURE;
  }
}

File: modules/common/ipc/examples/py/sub_example.py
# =================================================================================================
# Copyright (C) 2025 GRAPE Contributors
# =================================================================================================

import time
from datetime import datetime
from grape_ipc_py import init, Config, RawSubscriber, ok


def from_bytes(data: bytes) -> str:
    """Convert bytes to a string."""
    return data.decode("utf-8")


def data_callback(sample):
    """Callback for received data."""
    message = from_bytes(sample.data())
    publish_time = sample.info.publish_time
    print(f"Received message: '{message}' at {publish_time}")


def match_callback(match):
    """Callback for match/unmatch events."""
    print(f"Match status: {match.status}")


def main():
    try:
        # Initialize the IPC session
        config = Config()
        config.name = "Python IPC subscriber example"
        init(config)

        # Define the topic
        topic = "hello_world"

        # Create the subscriber
        subscriber = RawSubscriber(topic, data_callback, match_callback)

        # Sleep loop to keep the subscriber running
        SLEEP_TIME = 0.5  # 500 milliseconds
        while ok():
            time.sleep(SLEEP_TIME)

    except Exception as e:
        print(f"An error occurred: {e}")


if __name__ == "__main__":
    main()
